add expressions (regex, concat, at, length)

stuartin opened the pull request:
Allows the ability to run a regex match against a string, useful for parsing existing properties such as css font strings. Also document the concat expression

Fixes https://github.com/openlayers/openlayers/issues/16998

bjornharrtell commented:
I've missed something like this several times. Seems very nice to me, thanks @stuartin. :)

jbo023 commented:
We really like the new options, is there anything we can do to help to get this merged ?

ahocevar commented:
if @stuartin feel like this is ready for review, they should mark it as ready for review. This would be a first step to get it merged. Conflicts should be resolved first though.

stuartin commented:
It's been a little while since I looked at this, but the delay was waiting on feedback on how to approach literal values with the at operator.

If one of the maintainers can provide some input on how to proceed I could look at picking this back up.

https://github.com/openlayers/openlayers/issues/16998#issuecomment-3177460283

ahocevar commented:
@stuartin If these two work:

* `['at', 2, ['a', 'b', 'c'] -> 'c'`
* `['at', 2, ['get', 'myArray']] -> 'c'`

with
```js
{
  'myArray': ['a', 'b', 'c']'
}
```
then I think we're good.

Is this the guidance you've been looking for?

jahow commented on src/ol/expr/cpu.js:
```diff
@@ -376,6 +379,81 @@ function compileLogicalExpression(expression, context) {
+        const array = args[1](context);
+        if (Array.isArray(array)) {
+          // an array from an expression
+          return array[index];
```
this will run also if the argument is an expression like `['get', 'myProp']` right?

stuartin replied:
correct, covered in this test case
https://github.com/openlayers/openlayers/pull/16999/changes#diff-1dde688ecb2a04b0d6507e2129feece06035264afc52558a2337dac184211cbeR830

jahow commented on test/node/ol/expr/cpu.test.js:
```diff
@@ -807,6 +807,70 @@ describe('ol/expr/cpu.js', () => {
+      {
+        name: 'at (with exact array)',
+        type: StringType,
+        expression: ['at', 0, ['one', 'two', 'three']],
```
I don't think this is right; we ask for a `literal` operator when the user gives an array of strings; otherwise we will always consider it to be an expression (since it's an array with a string as first element)

stuartin replied:
removed the ability for `at` to work with arrays, it only works with `literal`, `get` or `regex` (assuming get would return an array)

jahow commented on test/node/ol/expr/cpu.test.js:
```diff
@@ -807,6 +807,70 @@ describe('ol/expr/cpu.js', () => {
+          },
+        },
+        type: StringType,
+        expression: ['at', 0, [['get', 'text'], 'two', 'three']],
```
same, I don't think we should support this; either the array is a literal array with fixed value, or we use an expression to compute it

stuartin replied:
doesnt support this any more, only `literal`, `get` or `regex`

jahow commented on test/node/ol/expr/expression.test.js:
```diff
@@ -471,6 +471,30 @@ describe('ol/expr/expression.js', () => {
+      },
+      {
+        name: 'index must be positive number (at)',
+        expression: ['at', -1, ['literal', ['one', 'two', 'three']]],
```
here I would suggest mimicking the `Array.at()` method from vanilla javascript, which allows negative indices: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at

stuartin replied:
added support for negative indexes using `Array.at()` (thanks for this!)

jahow commented on src/ol/expr/expression.js:
```diff
@@ -1071,6 +1093,91 @@ function withInArgs(encoded, returnType, context) {
+
+  if (!['literal', 'get', 'regex'].includes(arr[0])) {
+    throw new Error(
+      `failed to parse "at" expression: only "literal", "get" and "regex" expressions are supported`,
```
an array of numbers would actually be ok, without the 'literal' wrapping operator. This operator is really there for string arrays, see: https://github.com/openlayers/openlayers/blob/48327e000510b5ddb6d69d785cddc51c5388c7c9/src/ol/expr/expression.js#L105-L109

Ideally these array-related operators should be as close as possible to the existing `in` operator

jahow replied:
(and by that I mean that in the documentation it should be clear to the user, e.g. they share the same limitations etc.)

stuartin replied:
updated the logic to allow number arrays, with some additional tests and more explicit documentation. Let me know if you dont want to support mixed literal values like `['literal' ['one', 2, 3]` (currently it does)

stuartin committed:
add expressions (regex, concat, at, length)

String expressions:
Allows the ability to run a 'regex' match against a string, useful for parsing existing properties such as css font strings. Also document the 'concat' expression

Lookup expressions:
Allows the ability to get the 'length' of an array or string, in addition to being able to get a specific array item 'at' an index
